scipy/signal/_multimethods.py

```diff
@@ -0,0 +1,574 @@
+import functools
+import numpy as np
+from scipy._lib.uarray import Dispatchable, all_of_type, create_multimethod
+from scipy.signal import _api
+from scipy.signal._backend import scalar_or_array, tuple_str_array
+
+__all__ = [
+    'upfirdn', 'sepfir2d',
+    # signaltools
+    'correlate', 'correlation_lags', 'correlate2d',
+    'convolve', 'convolve2d', 'fftconvolve', 'oaconvolve',
+    'order_filter', 'medfilt', 'medfilt2d', 'wiener', 'lfilter',
+    'lfiltic', 'sosfilt', 'deconvolve', 'hilbert', 'hilbert2',
+    'cmplx_sort', 'unique_roots', 'invres', 'invresz', 'residue',
+    'residuez', 'resample', 'resample_poly', 'detrend',
+    'lfilter_zi', 'sosfilt_zi', 'sosfiltfilt', 'choose_conv_method',
+    'filtfilt', 'decimate', 'vectorstrength',
+    # waveforms
+    'sawtooth', 'square', 'gausspulse', 'chirp', 'sweep_poly',
+    'unit_impulse'
+]
+
+
+_create_signal = functools.partial(
+                    create_multimethod,
+                    domain="numpy.scipy.signal"
+                 )
+
+_mark_scalar_or_array = functools.partial(
+                            Dispatchable,
+                            dispatch_type=scalar_or_array,
+                            coercible=True
+                        )
+
+_mark_tuple_str_array = functools.partial(
+                            Dispatchable,
+                            dispatch_type=tuple_str_array,
+                            coercible=True
+                        )
+
+
+def _get_docs(func):
+    func.__doc__ = getattr(_api, func.__name__).__doc__
+    return func
+
+
+def _h_x_replacer(args, kwargs, dispatchables):
+    def self_method(h, x, *args, **kwargs):
+        return (dispatchables[0], dispatchables[1]) + args, kwargs
+
+    return self_method(*args, **kwargs)
+
+
+@_create_signal(_h_x_replacer)
+@all_of_type(np.ndarray)
+@_get_docs
+def upfirdn(h, x, up=1, down=1, axis=-1, mode='constant', cval=0):
+    return h, x
+
+
+def _input_hrow_hcol_replacer(args, kwargs, dispatchables):
+    def self_method(input, hrow, hcol, *args, **kwargs):
+        return (dispatchables[0], dispatchables[1],
+                dispatchables[2]) + args, kwargs
+
+    return self_method(*args, **kwargs)
+
+
+@_create_signal(_input_hrow_hcol_replacer)
+@all_of_type(np.ndarray)
+@_get_docs
+def sepfir2d(input, hrow, hcol):
+    return input, hrow, hcol
```

Comment:
Can we use a code generator to produce `_multimethods.py` file? Like a yml file which can contain all the information. We just need to define the replacer methods and these kind of 3-4 liner blocks can be generated dynamically? cc: @grlee77 @rgommers @AnirudhDagar 

P.S. It might not be worth it because anyways its the same amount of code in both cases (auto generation and manually writing). But still thought to put this up.

Reply:
I know that sounds appealing and probably better engineered (will help with consistency) but I believe there are not many uarray modules and multimethods left to be added now (with the PRs already open) and autogeneration might just make things more complicated (one more thing to be handled).